### ISSUE TITLE
OS X testing of unusual filenames (non-ascii)

### ISSUE THREAD

**Issue by ThomasWaldmann** [borgbackup/borg#143]
See: borg.testsuite.archiver.ArchiverTestCase.test_unusual_filenames

Currently tests filenames with blanks and with parens.
Initially it also tested a filenames with non-ascii characters.
The non-ascii filename worked on linux, but failed on OS X (on travis-ci).

As it is hard to debug such stuff on travis, I removed the non-ascii filename again, but this should be checked by a developer with an OS X machine.

If a non-developer wants to help, try this in a OS X terminal:
```
borg init repo
mkdir input
touch input/non_ascii_filename_äöüß
borg create repo::test input
mkdir output
cd output
borg extract ../repo::test input/non_ascii_filename_äöüß
ls -l input
```
Is the extracted file there?

**Comment by mhubig** [comment]
```bash
~ ❯❯❯ pip3 install borgbackup
You are using pip version 7.1.0, however version 7.1.2 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
Collecting borgbackup
  Downloading borgbackup-0.24.0.tar.gz (276kB)
    100% |████████████████████████████████| 278kB 331kB/s
Collecting msgpack-python>=0.4.6 (from borgbackup)
  Downloading msgpack-python-0.4.6.tar.gz (116kB)
    100% |████████████████████████████████| 118kB 387kB/s
Building wheels for collected packages: borgbackup, msgpack-python
  Running setup.py bdist_wheel for borgbackup
  Stored in directory: /Users/markus/Library/Caches/pip/wheels/d6/e1/3c/e69fe3efe54b9724f728d9ed47f91db2b3840af22350e13784
  Running setup.py bdist_wheel for msgpack-python
  Stored in directory: /Users/markus/Library/Caches/pip/wheels/f3/97/a5/dd6e3b680de10b689464c44bc211239d1fe54bd296ff860897
Successfully built borgbackup msgpack-python
Installing collected packages: msgpack-python, borgbackup
Successfully installed borgbackup-0.24.0 msgpack-python-0.4.6
~/tmp ❯❯❯ borg init repo
Initializing repository at "repo"
Encryption NOT enabled.
Use the "--encryption=repokey|keyfile|passphrase" to enable encryption.
Synchronizing chunks cache...
Rebuilding archive collection. Known: 0 Repo: 0 Unknown: 0
Merging collection into master chunks cache...
Done.
~/tmp ❯❯❯ mkdir input
~/tmp ❯❯❯ touch input/non_ascii_filename_äöüß
~/tmp ❯❯❯ borg create repo::test input
~/tmp ❯❯❯ mkdir output
~/tmp ❯❯❯ cd output
~/t/output ❯❯❯ borg extract ../repo::test input/non_ascii_filename_äöüß
~/t/output ❯❯❯ ls
~/t/output ❯❯❯ ls -l input
ls: input: No such file or directory
```

**Comment by mhubig** [comment]
```bash
~/Development ❯❯❯ git clone https://github.com/borgbackup/borg.git
Cloning into 'borg'...
remote: Counting objects: 5690, done.
remote: Total 5690 (delta 0), reused 0 (delta 0), pack-reused 5690
Receiving objects: 100% (5690/5690), 1.54 MiB | 932.00 KiB/s, done.
Resolving deltas: 100% (3871/3871), done.
Checking connectivity... done.
~/Development ❯❯❯ cd borg
~/D/borg git:master ❯❯❯ git describe
0.24.0-57-g034e347
mkvirtualenv -a $(pwd) -p python3 borg
Running virtualenv with interpreter /usr/local/bin/python3
Using base prefix '/usr/local/Cellar/python3/3.4.3_2/Frameworks/Python.framework/Versions/3.4'
New python executable in borg/bin/python3.4
Also creating executable in borg/bin/python
Installing setuptools, pip, wheel...done.
Setting project for borg to /Users/markus/Development/borg
(borg)~/D/borg git:master ❯❯❯ pip install cython                                                          ⏎
Collecting cython
  Downloading Cython-0.23.1-cp34-cp34m-macosx_10_6_intel.macosx_10_9_intel.macosx_10_9_x86_64.macosx_10_10_intel.macosx_10_10_x86_64.whl (4.5MB)
    100% |████████████████████████████████| 4.5MB 113kB/s
Installing collected packages: cython
Successfully installed cython-0.23.1
(borg)~/D/borg git:master ❯❯❯ brew install lz4                                                            ⏎
Warning: You are using OS X 10.11.
We do not provide support for this pre-release version.
You may encounter build failures or other breakage.
==> Downloading https://github.com/Cyan4973/lz4/archive/r131.tar.gz
==> Downloading from https://codeload.github.com/Cyan4973/lz4/tar.gz/r131
######################################################################## 100,0%
==> make install PREFIX=/usr/local/Cellar/lz4/r131
🍺  /usr/local/Cellar/lz4/r131: 12 files, 432K, built in 14 seconds
(borg)~/D/borg git:master ❯❯❯ python setup.py install
....
Finished processing dependencies for borgbackup==0.24.0+57.g034e347
(borg)~/D/borg git:master ❯❯❯ mkdir tmp
(borg)~/D/borg git:master ❯❯❯ cd tmp
(borg)~/D/b/tmp git:master ❯❯❯ borg init repo
(borg)~/D/b/tmp git:master ❯❯❯ mkdir input                                                                ⏎
(borg)~/D/b/tmp git:master ❯❯❯ touch input/non_ascii_filename_äöüß
(borg)~/D/b/tmp git:master ❯❯❯ borg create repo::test input                                               
(borg)~/D/b/tmp git:master ❯❯❯ mkdir output                                                               
(borg)~/D/b/tmp git:master ❯❯❯ cd output                                                                  
(borg)~/D/b/t/output git:master ❯❯❯ borg extract ../repo::test input/non_ascii_filename_äöüß              
(borg)~/D/b/t/output git:master ❯❯❯ ls -l input                                                           
ls: input: No such file or directory
````

**Comment by mhubig** [comment]
````bash
borg)~/D/borg git:master ❯❯❯ git checkout experimental
Branch experimental set up to track remote branch experimental from origin.
Switched to a new branch 'experimental'
(borg)~/D/borg git:experimental ❯❯❯ pip install blosc
(borg)~/D/borg git:experimental ❯❯❯ python setup.py install
....
Finished processing dependencies for borgbackup==0.16+305.g3a38457
(borg)~/D/borg git:experimental ❯❯❯ mkdir tmp                                                             
(borg)~/D/borg git:experimental ❯❯❯ cd tmp                                                                
(borg)~/D/b/tmp git:experimental ❯❯❯ borg init repo                                                       
Initializing repository at "repo"
Encryption NOT enabled.
Use the "--encryption=passphrase|keyfile" to enable encryption.
Synchronizing chunks cache...
Rebuilding archive collection. Known: 0 Repo: 0 Unknown: 0
Merging collection into master chunks cache...
Done.
(borg)~/D/b/tmp git:experimental ❯❯❯ mkdir input                                                          
(borg)~/D/b/tmp git:experimental ❯❯❯ touch input/non_ascii_filename_äöüß                                  
(borg)~/D/b/tmp git:experimental ❯❯❯ borg create repo::test input                                         
(borg)~/D/b/tmp git:experimental ❯❯❯ mkdir output                                                         
(borg)~/D/b/tmp git:experimental ❯❯❯ cd output                                                            
(borg)~/D/b/t/output git:experimental ❯❯❯ borg extract ../repo::test input/non_ascii_filename_äöüß        
(borg)~/D/b/t/output git:experimental ❯❯❯ ls -l input                                                     
ls: input: No such file or directory
````

**Comment by mhubig** [comment]
:-(

**Comment by ThomasWaldmann** [comment]
Thanks for testing. So we know now that it behaves as on travis.

Now this needs debugging - why does it happen?

**Comment by mhubig** [comment]
````zsh
(borg)~/D/b/b/testsuite git:master ❯❯❯ mkdir tmp
(borg)~/D/b/b/testsuite git:master ❯❯❯ cd tmp
(borg)~/D/b/b/t/tmp git:master ❯❯❯ borg init repo
Initializing repository at "repo"
Encryption NOT enabled.
Use the "--encryption=repokey|keyfile|passphrase" to enable encryption.
Synchronizing chunks cache...
Rebuilding archive collection. Known: 0 Repo: 0 Unknown: 0
Merging collection into master chunks cache...
Done.
(borg)~/D/b/b/t/tmp git:master ❯❯❯ mkdir input                                                            
(borg)~/D/b/b/t/tmp git:master ❯❯❯ touch input/non_ascii_filename_äöüß                                    
(borg)~/D/b/b/t/tmp git:master ❯❯❯ borg create repo::test input                                           
(borg)~/D/b/b/t/tmp git:master ❯❯❯ borg list repo                                                         
test                                 Wed Aug 26 15:18:09 2015
(borg)~/D/b/b/t/tmp git:master ❯❯❯ borg list repo::test                                                   
drwxr-xr-x markus staff         0 Aug 26 15:18 input
-rw-r--r-- markus staff         0 Aug 26 15:18 input/non_ascii_filename_äöüß
````

Since the `list` command shows the file, I would assume the problem is within the `extract` routine ...

**Comment by mhubig** [comment]
````bash
(borg)~/D/b/b/t/tmp git:master ❯❯❯ mkdir output                                                           
(borg)~/D/b/b/t/tmp git:master ❯❯❯ cd output                                                              
(borg)~/D/b/b/t/t/output git:master ❯❯❯ borg extract ../repo::test
(borg)~/D/b/b/t/t/output git:master ❯❯❯ ls input/non_ascii_filename_äöüß                                  
input/non_ascii_filename_äöüß
(borg)~/D/b/b/t/t/output git:master ❯❯❯ rm -rf input
(borg)~/D/b/b/t/t/output git:master ❯❯❯ borg extract ../repo::test2 input                                 
(borg)~/D/b/b/t/t/output git:master ❯❯❯ ls input                                                          
non_ascii_filename_äöüß
````
It seems the problem occurs only If I give the strange filename as argument ...

**Comment by ThomasWaldmann** [comment]
Here's some minimal code to test argparsing:

```
import sys
import argparse

print(repr(sys.argv))

parser = argparse.ArgumentParser(add_help=False)
parser.add_argument('--path', dest='path', metavar='PATH', help='set path')
args = parser.parse_args(sys.argv[1:])
print(repr(args))
```

Result on Linux:
```
$ python3.4 args.py --path some_path_äöü
['args.py', '--path', 'some_path_äöü']
Namespace(path='some_path_äöü')
```

Make sure you run it with python 3.x (not: 2.x).

**Comment by mhubig** [comment]
I could further dig down the problem to `borg.helpers.exclude_path` and `borg.helpers.IncludePattern`:

````python
def exclude_path(path, patterns):
    """Used by create and extract sub-commands to determine
    whether or not an item should be processed.
    """
    print(path, type(path))
    print(patterns)
    for pattern in (patterns or []):
        print(pattern)
        print('Match:', pattern.match(path), pattern, path)
        if pattern.match(path):
            print('match')
            print('excludepattern:', isinstance(pattern, ExcludePattern))
            return isinstance(pattern, ExcludePattern)
    return False

class IncludePattern:
    """Literal files or directories listed on the command line
    for some operations (e.g. extract, but not create).
    If a directory is specified, all paths that start with that
    path match as well.  A trailing slash makes no difference.
    """
    def __init__(self, pattern):
        self.pattern = os.path.normpath(pattern).rstrip(os.path.sep)+os.path.sep

    def match(self, path):
        print("path:", type(path), path)
        print("path+os.path.sep:", type(path+os.path.sep), path+os.path.sep)
        print("pattern:", type(self.pattern), self.pattern)
        print("(path+os.path.sep).startswith(self.pattern):", (path+os.path.sep).startswith(self.pattern))
        return (path+os.path.sep).startswith(self.pattern)

    def __repr__(self):
        return '%s(%s)' % (type(self), self.pattern)
````

Running `borg extract ../repo::test2 input/non_ascii_filename_äöüß` with the annotations above gives me this output:

````bash
(borg)~/D/b/t/output git:master ❯❯❯ borg extract ../repo::test2 input/non_ascii_filename_äöüß                                                            
input <class 'str'>
[<class 'borg.helpers.IncludePattern'>(input/non_ascii_filename_äöüß/), <class 'borg.helpers.ExcludePattern'>(*/*)]
<class 'borg.helpers.IncludePattern'>(input/non_ascii_filename_äöüß/)
path: <class 'str'> input
path+os.path.sep: <class 'str'> input/
pattern: <class 'str'> input/non_ascii_filename_äöüß/
(path+os.path.sep).startswith(self.pattern): False
Match: False <class 'borg.helpers.IncludePattern'>(input/non_ascii_filename_äöüß/) input
path: <class 'str'> input
path+os.path.sep: <class 'str'> input/
pattern: <class 'str'> input/non_ascii_filename_äöüß/
(path+os.path.sep).startswith(self.pattern): False
<class 'borg.helpers.ExcludePattern'>(*/*)
Match: True <class 'borg.helpers.ExcludePattern'>(*/*) input
match
excludepattern: True
input/non_ascii_filename_äöüß <class 'str'>
[<class 'borg.helpers.IncludePattern'>(input/non_ascii_filename_äöüß/), <class 'borg.helpers.ExcludePattern'>(*/*)]
<class 'borg.helpers.IncludePattern'>(input/non_ascii_filename_äöüß/)
path: <class 'str'> input/non_ascii_filename_äöüß
path+os.path.sep: <class 'str'> input/non_ascii_filename_äöüß/
pattern: <class 'str'> input/non_ascii_filename_äöüß/
(path+os.path.sep).startswith(self.pattern): False
Match: False <class 'borg.helpers.IncludePattern'>(input/non_ascii_filename_äöüß/) input/non_ascii_filename_äöüß
path: <class 'str'> input/non_ascii_filename_äöüß
path+os.path.sep: <class 'str'> input/non_ascii_filename_äöüß/
pattern: <class 'str'> input/non_ascii_filename_äöüß/
(path+os.path.sep).startswith(self.pattern): False
<class 'borg.helpers.ExcludePattern'>(*/*)
Match: True <class 'borg.helpers.ExcludePattern'>(*/*) input/non_ascii_filename_äöüß
match
excludepattern: True
(borg)~/D/b/t/output git:master ❯❯❯
````

The Problem is, that `.startwith()` is somehow **NOT** returning `True` here! But if I put this into the python3 repl it works ...

````python
>>> import os
>>> path = 'input/non_ascii_filename_äöüß'
>>> pattern = 'input/non_ascii_filename_äöüß/'
>>> (path+os.path.sep).startswith(pattern)
True
````

This strange behavior leeds to the problem that the path to extract is filtered away within `borg.archive.DownloadPipeline.unpack_many`:

````python
def unpack_many(self, ids, filter=None, preload=False):
        unpacker = msgpack.Unpacker(use_list=False)
        for data in self.fetch_many(ids):
            unpacker.feed(data)
            items = [decode_dict(item, (b'path', b'source', b'user', b'group')) for item in unpacker]
            if filter:
                print("PreFilter: {}".format(items))
                items = [item for item in items if filter(item)]
                print("PostFilter: {}".format(items))
            if preload:
                for item in items:
                    if b'chunks' in item:
                        self.repository.preload([c[0] for c in item[b'chunks']])
            for item in items:
                yield item
````

The annotated code above gives the following output with an non ascii filename:

````bash
(borg)~/D/b/t/output git:master ❯❯❯ borg extract ../repo::test2 input/non_ascii_filename_äöüß                                                            
PreFilter: [{b'gid': 20, b'uid': 501, b'mode': 16877, b'mtime': 1440595083000000000, b'path': 'input', b'group': 'staff', b'user': 'markus'}, {b'chunks': ((b'\x13\xb8\x96\xd5Q\xa1\x00@\x1b\r9\x82\xe0r\x9e\xfc.\x8dz\xeb\t\xa3l\nQ\xe4\x8e\xc2\xbd\x15\xea\x8b', 5, 8),), b'gid': 20, b'mode': 33188, b'mtime': 1440595419000000000, b'uid': 501, b'path': 'input/non_ascii_filename_äöüß', b'group': 'staff', b'user': 'markus'}]
PostFilter: []
(borg)~/D/b/t/output git:master ❯❯❯
````

And with an ascii only filename:

````bash
(borg)~/D/b/t/output git:master ❯❯❯ borg extract ../repo::test3 input/plain_ascii_filename                                                               
PreFilter: [{b'mode': 33188, b'mtime': 1440671262000000000, b'chunks': ((b'\x1d7+\xced\xd9\xfeL=\xdf\xc9\r+(L\xfc\xaf\x1dt\xb9Z!\xfd\xf6\xa4\x04[\x00\x95\x82\xe1D', 13, 16),), b'uid': 501, b'group': 'staff', b'user': 'markus', b'gid': 20, b'path': 'input/plain_ascii_filename'}]
PostFilter: [{b'mode': 33188, b'mtime': 1440671262000000000, b'chunks': ((b'\x1d7+\xced\xd9\xfeL=\xdf\xc9\r+(L\xfc\xaf\x1dt\xb9Z!\xfd\xf6\xa4\x04[\x00\x95\x82\xe1D', 13, 16),), b'uid': 501, b'group': 'staff', b'user': 'markus', b'gid': 20, b'path': 'input/plain_ascii_filename'}]
(borg)~/D/b/t/output git:master ❯❯❯
````

**Comment by mhubig** [comment]
As a next step I'll try to replace the `.startswith()` call with `regex.match()` maybe this will be a workaround for now ... but the real cause of the problem is not at all clear for me. :-(

**Comment by edgewood** [comment]
Your message tickled a memory, so I did a bit of searching.  The problem 
is that Unicode allows a few different ways of forming some characters.  
Consider 'ä'.  That could be formed by using the single Unicode 
character LATIN SMALL LETTER A WITH DIAERESIS.

  >>> composed = "\N{LATIN SMALL LETTER A WITH DIAERESIS}"
  >>> composed.encode()
  b'\xc3\xa4'

This is known as a composed form.  But it could also be formed by 
appending a COMBINING DIAERESIS to a LATIN SMALL LETTER A.

  >>> decomposed = "\N{LATIN SMALL LETTER A}\N{COMBINING DIAERESIS}"
  >>> decomposed.encode()
  b'a\xcc\x88'

That's a decomposed form.

OS X is unusual among Unix(ish) operating systems in that it specifies 
that Unicode characters are (mostly) decomposed before being written as 
HFS+ filenames.  So if you say "create a file named ä", the bytes of the 
file name will always be b'a\xcc\x88', even if you typed b'\xc3\xa4'.

The problem here is that:

  >>> decomposed.startswith(composed)
  False

I'll bet if you .encode() the strings, you'll find that the path from 
borg's repository is in decomposed form, but what you typed on the 
command line is in composed form.  And if you type the filename on the 
command line in decomposed form, borg will find it.

A solution (not necessarily the best, perhaps) would be to import 
unicodedata, then convert a copy of the command line path and a copy of 
the repository path to a common form before comparing them.

  >>> import unicodedata as ud
  >>> ud.normalize("NFC", decomposed).startswith(ud.normalize("NFC", composed))
  True

This link talks about the exceptions where HFS+ *doesn't* decompose 
certain characters: http://apple.stackexchange.com/a/84038

I don't think that would affect my solution, but might affect alternate 
solutions, and in any case the exceptions would be good unit test cases.

**Comment by mhubig** [comment]
Indeed normalizing the two strings before comparing them fixes the problem!

````python
def match(self, path):
    import unicodedata as ud
    path = ud.normalize('NFC', path+os.path.sep)
    pattern = ud.normalize('NFC', self.pattern)
    return path.startswith(pattern)
````
````bash
❯❯❯ borg extract ../repo::test2 input/non_ascii_filename_äöüß           
❯❯❯ ls input/non_ascii_filename_äöüß                                    
input/non_ascii_filename_äöüß
````
Now the question remains where the normalization should be done? Obviously doing it within the `match` function would result in doing it for every path within the archive, every time we're extracting something. It seems the *decomposed string* comes from the archive and the *composed string* from the command line. So maybe it would be a good idea to 'normalize' the pathnames before storing them in the archive and also normalize the paths from the command line before using them?!

Or maybe it would be sufficient to somehow transcode the paths we get from the command line to the decomposed form (but only we are on a mac, or maybe only if we are extracting from an archive made on a mac?) ...

I could try to provide a PR, but it would be nice to get some opinions first ...

PS.: Very well @edgewood ... you must have been **deep down** that unicode dungeon my friend! :-)

**Comment by ThomasWaldmann** [comment]
@mhubig @edgewood good debugging / bug hunting, thanks!

There is a problem with treating path/filenames (in many ways, including "normalization") before storing them into the archive: IIRC, a unix path/filename can be any zero-terminated byte sequence (including rather nonsensical ones, like "funny characters" that came from wrong encodings, often seen on samba shares).

Thus, trying to do any byte sequence -> unicode decoding and normalization is likely to cause issues.
We can not rely on getting input in a specific encoding from the filesystem, there can be wild mixups and broken stuff there.

Also, I think we really should store the original path/filename byte sequence, otherwise we lose information about it.

A PR would be welcome. Please extend the test I mentioned above so it includes the non-ascii stuff.

Also extend tests to test with really broken filenames (like ISO-8859-1 interpreted as UTF-8 or so), so we will see if the .normalize() blows up then.

**Comment by ThomasWaldmann** [comment]
BTW, as 0.25 will be out very soon, I've added a FAQ entry for now. A real fix is still needed, of course.

**Comment by edgewood** [comment]
That's my understanding as well: Unix filesystems don't impose any particular encoding, and files can contain byte sequences that don't belong to any particular encoding.  Encoding is left to the user interface, and backup utilities shouldn't change them.

The two strategies (adapted from @mhubig) that I think would work are to convert command line paths to a OSX decomposed form when the command line is being parsed, or to change the *Pattern classes so that they can normalize both the pattern and the path when appropriate (just on OSX?), and not otherwise.

Converting the command line paths probably has the lowest runtime cost, since there are usually a very small number of command line paths compared to the number of paths in an archive.

The disadvantage is that HFS+ paths are not precisely NFD, but have a number of exceptions.  We'd need code that would normalize to the OSX variant of NFD.  I don't know how to write that code, and don't have access to OSX to test it.

Changing the *Pattern classes to be able to normalize both patterns and paths when appropriate, and not when not, probably shouldn't be too big of a performance hit when needed: using timeit to convert mostly NFD form strings to NFD showed that one million loops took about a half second on my system.  If done correctly, the "no normalization" path should have negligable impact if done right.

I've created a standalone test case at <http://pastebin.com/GcfrHBeb>.  Comments welcome, as I don't have a whole lot of Python dev experience and am still figuring out best practices.  If @ThomasWaldmann wants to go that route, I'll update ExcludePattern similarly and create legit unit test cases, then create a PR.

-- 
Ed Blackman

**Comment by ThomasWaldmann** [comment]
@edgewood if you could either use a PR or codereview.appspot.com ("before commit"), that would help a lot. Manually comparing a pastebin to current code is not that comfortable.

**Comment by mhubig** [comment]
@edgewood: Have you thought about these corner cases?

* Extracting an archive made on Linux with Linux: Keeps Working!
* Extracting an archive made on Linux with OS X: Breaks?
* Extracting an archive made on OS X with Linux. Is and will be Broken?
* Extracting an archive made on OS X with OS X. Will be fixed!

I totally agree that backup utilities shouldn't change the encoding! But maybe we could store some encoding information within the archive, so we can use the right one when accessing/extracting the archive?

**Comment by edgewood** [comment]
Pull request created.  I am still setting up all the requirements to run the testsuite, need to go, and might not get a chance to work on this for a few days.  So the new test cases might have a bug or two!

-- 
Ed Blackman

**Comment by mhubig** [comment]
link #171

**Comment by ThomasWaldmann** [comment]
@mhubig good idea to think about cross-platform restore behaviour.

the problem with "storing the encoding" is that (at least on non-OSX) we can't know, it is just bytestring, they don't need to have one specific encoding (or even a valid encoding).

**Comment by ThomasWaldmann** [comment]
btw, a really simple workaround might be to replace non-ascii stuff by wildcards on the commandline.
that could work everywhere.

**Comment by ThomasWaldmann** [comment]
Note: I've put this into next milestone, but there needs to be some more work before it can be merged:

- have 1 pull request about it, not 2
- a mechanism that works for osx as well as for other platforms

**Comment by edgewood** [comment]
Sorry, have had less free time than expected. Will work on the feedback this weekend.
-- 
Ed Blackman
Sent from my OnePlus One. Please excuse my brevity.


-------- Original Message --------

**Comment by ThomasWaldmann** [comment]
should be fixed by PR #193, now merged into master. thanks for analyzing/fixing this!
